### PR TITLE
Remove the border when draw has no title

### LINKED ISSUE #14178: There is an extra blank headers at the top of drawer component
- [ ] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[https://ant-design.gitee.io/components/drawer-cn/](https://ant-design.gitee.io/components/drawer-cn/)

### Steps to reproduce
升级到最新版即可

### What is expected?
希望能做到向后兼容，默认不显示，在用户设置了title属性时再显示

### What is actually happening?
在未设置title属性时，drawer顶部会显示一个多余的空白行

| Environment | Info |
|---|---|
| antd | 3.12.1 |
| React | 16.6.3 |
| System | mac 10.14.1  |
| Browser | chrome 71.0.3578.98 |

---
drawer的关闭按钮之前是和ant-drawer-body同级的节点，现在被放到了ant-drawer-header(最新版新增)中了，这会导致页面多出来一个空白头

### PR DESCRIPTION
**Pull request by ztplz**
First of all, thank you for your contribution! 😄

New feature please send pull request to feature branch, and rest to master branch.
Pull request will be merged after one of collaborators approve.
Please makes sure that these form are filled before submitting your pull request, thank you!

[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / document update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Branch merge
- [ ] Other (about what?)

### What's the background?

> 1. Describe the source of requirement.
> 2. Resolve what problem.
Try to fix https://github.com/ant-design/ant-design/issues/14178
> 3. Related issue link.
  
### API Realization (Optional if not new feature)

> 1. Basic thought of solution and other optional proposal.
> 2. List final API realization and usage sample.
> 3. GIF or snapshot should be provided if includes UI/interactive modification.

### What's the effect? (Optional if not new feature)

> 1. Does this PR affect user? Which part will be affected?
> 2. What will say in changelog?
> 3. Does this PR contains potential break change or other risk?

### Changelog description (Optional if not new feature)

> 1. English description
> 2. Chinese description (optional)

### Self Check before Merge

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### Additional Plan? (Optional if not new feature)

> If this PR related with other PR or following info. You can type here.

### CODE REVIEW

File: components/drawer/style/drawer.less [hunk 1]
```diff
@@ -153,6 +153,13 @@
     position: relative;
   }
 
+  &-header-no-title {
+    padding: 16px 24px;
```
Comment (afc163) [inline components/drawer/style/drawer.less]: 这个 padding 也不应该留，不然会挤压内容，留下一行空行。

Comment (ztplz) [reply components/drawer/style/drawer.less]: 可以减少一点 上下8px感觉就够了

File: components/drawer/index.tsx [hunk 2]
```diff
@@ -156,11 +156,13 @@ class Drawer extends React.Component<DrawerProps & ConfigConsumerProps, IDrawerS
 
   renderHeader() {
     const { title, prefixCls, closable } = this.props;
-    if (!title && !closable) {
+    if (!closable) {
       return null;
```
Comment (dengfuping) [inline components/drawer/index.tsx]: `title` 为空应该返回 `null`，原先的 `!closable` 判断应该有误。

File: components/drawer/index.tsx [hunk 3]
```diff
@@ -156,11 +156,13 @@ class Drawer extends React.Component<DrawerProps & ConfigConsumerProps, IDrawerS
...
     }
+
+    const headerClassName = title ? `${prefixCls}-header` : `${prefixCls}-header-no-title`;
     return (
-      <div className={`${prefixCls}-header`}>
+      <div className={headerClassName}>
         {title && <div className={`${prefixCls}-title`}>{title}</div>}
         {this.renderCloseIcon()}
```
Comment (dengfuping) [inline components/drawer/index.tsx]: - `title` 为空时不应该渲染 `header`
-  `closeIcon` 应该渲染到 `body` 的位置

File: components/drawer/style/drawer.less [hunk 4]
```diff
@@ -153,6 +153,12 @@
     position: relative;
   }
 
+  &-header-no-title {
+    background: @component-background;
+    color: @text-color;
+    position: relative;
+  }
+
```
Comment (dengfuping) [inline components/drawer/style/drawer.less]: 这个样式也就不需要了

File: components/drawer/index.tsx (hunk 3 above)
Comment (afc163) [reply components/drawer/index.tsx]: 放在一起是为了解决 https://github.com/ant-design/ant-design/pull/14030

Comment (ztplz) [reply components/drawer/index.tsx]: 这样就要处理情况
1.有title，closeable，border 和 closeIcon 都要有
2.只有title，有border 没 closeIcon
3.只有closeable, 没 border 有 closeIcon
4. 都没有

Comment (ztplz) [reply components/drawer/index.tsx]: 这样就需要几种情况
1.title，closeable 都有，header 和 closeIcon 都要有
2.只有title，那就只有header
3.只有closeable， 那就只有 closeIcon
4.都没有

Comment (zombieJ) [reply components/drawer/index.tsx]: 感觉这样就复杂了，要不然拿出来？

Comment (ztplz) [reply components/drawer/index.tsx]: 拿出来只是位置不一样而已   还是要分几种情况的

Comment (chenshuai2144) [reply components/drawer/index.tsx]: 还是在里面好。close 应该是title 的一部分吧，拿出来应该叫 handle 之类的

Comment (dengfuping) [reply components/drawer/index.tsx]: > closeIcon 应该渲染到 body 的位置
> 放在一起是为了解决 #14030

- 这里的 body 指的是 Drawer 的 body 部分，即应该和 `Modal` 的行为保持一致，title 为空时不渲染 Drawer header, 关闭图片渲染到 Drawer 的 body 中。

image

image

Comment (dengfuping) [reply components/drawer/index.tsx]: > 还是在里面好。close 应该是title 的一部分吧，拿出来应该叫 handle 之类的

赞同 @chenshuai2144 说的，`closeIcon` 放在外面可以作为一个 `feature` 单独实现。

Comment (zombieJ) [reply components/drawer/index.tsx]: 不用作为 feature，还是放里面

Comment (ztplz) [reply components/drawer/index.tsx]: 没必要放外面做feature  用户想放外面可以自己在外面加  api太多了不好

**Comment by dengfuping** [comment]
@ztplz `commit` 包含了多个组件代码的变更，拆分成多个 `PR`?

**Comment by afc163** [comment]
混合了其他提交，Rebase 或者重提一个。

**Pull request closed without merging**